libraries/chain/include/eosio/chain/finality/qc.hpp

```diff
@@ -0,0 +1,262 @@
+
+   using vote_bitset   = fc::dynamic_bitset;
+   using bls_key_map_t = std::map<bls_public_key, bls_private_key>;
+
+   constexpr std::array weak_bls_sig_postfix = { 'W', 'E', 'A', 'K' };
+   using weak_digest_t = std::array<uint8_t, sizeof(digest_type) + weak_bls_sig_postfix.size()>;
+
+   inline weak_digest_t create_weak_digest(const digest_type& digest) {
+      weak_digest_t res;
+      std::memcpy(res.begin(), digest.data(), digest.data_size());
+      std::memcpy(res.begin() + digest.data_size(), weak_bls_sig_postfix.data(), weak_bls_sig_postfix.size());
+      return res;
+   }
+
+   enum class vote_status {
+      success,
+      duplicate,             // duplicate vote, expected as votes arrive on multiple connections
+      unknown_public_key,    // public key is invalid, indicates invalid vote
+      invalid_signature,     // signature is invalid, indicates invalid vote
+      unknown_block,         // block not available, possibly less than LIB, or too far in the future
+      max_exceeded           // received too many votes for a connection
+   };
+
+   enum class has_vote_status_t {
+      voted,
+      not_voted,
+      irrelevant_finalizer
+   };
+
+   struct qc_sig_t {
+      bool is_weak()   const { return !!weak_votes; }
+      bool is_strong() const { return !weak_votes; }
+
+      std::optional<vote_bitset> strong_votes;
+      std::optional<vote_bitset> weak_votes;
+      bls_aggregate_signature    sig;
+
+      // called from net threads
+      void verify(const finalizer_policy_ptr& fin_policy, const digest_type& strong_digest, const weak_digest_t& weak_digest) const;
+   };
+
+   struct qc_t {
+      uint32_t                  block_num{0};
+      // signatures corresponding to the active finalizer policy
+      qc_sig_t                  active_policy_sig;
+      // signatures corresponding to the pending finalizer policy if there is one
+      std::optional<qc_sig_t>   pending_policy_sig;
+
+      bool is_strong() const {
+         return active_policy_sig.is_strong() && (!pending_policy_sig || pending_policy_sig->is_strong());
+      }
+      bool is_weak() const { return !is_strong(); }
+
+      qc_claim_t to_qc_claim() const {
+         return { .block_num = block_num, .is_strong_qc = is_strong() };
+      }
+   };
+
+   struct qc_data_t {
+      std::optional<qc_t> qc;  // Comes either from traversing branch from parent and calling get_best_qc()
+                               // or from an incoming block extension.
+      qc_claim_t qc_claim;     // describes the above qc_t. In rare cases (bootstrap, starting from snapshot,
+                               // disaster recovery), we may not have a qc_t so we use the `lib` block_num
+                               // and specify `weak`.
+   };
+
+   /**
+    * All public methods are thread-safe.
+    * Used for incorporating votes into a qc signature.
+    * "open" in that it allows new votes to be added at any time.
+    */
+   class open_qc_sig_t {
```

Comment:
I feel `pending_qc_sig_t` and `pending_qc` are more appropriate. `pending` means something in progress; that's what they are. No need to worry about it might be confused with pending finalizer policy. We use pending in other cases in the code base too.

Reply:
Renamed to `in_progress_qc_sig_t` and `in_progress_qc_t`.

Reply:
Renamed to `aggregating_qc_sig_t` and `aggregating_qc_t`.